opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java

```diff
@@ -0,0 +1,412 @@
+      start += chunk.length;
+      ngrammer.reset(chunk);
+
+      while (ngrammer.hasNext()) {
+        String nGram = ngrammer.next();
+        if (nGram.equals(SPACE)) {
+          continue;
+        }
+        MutableInt cnt = ngramCounts.get(nGram);
+        if (cnt == null) {
+          ngramCounts.put(nGram, new MutableInt(1));
+        } else {
+          cnt.increment();
+        }
+        if (++nGrams % 110 == 0) {
+          currPredictions = predict(ngramCounts);
+          if (seenEnough(predictions, currPredictions, ngramCounts)) {
+            return currPredictions;
+          }
+        }
+      }
+    }
+  }
+
+  private Language[] predict(Map<String, MutableInt> ngramCounts) {
+    String[] allGrams = new String[ngramCounts.size()];
+    float[] counts = new float[ngramCounts.size()];
+    int i = 0;
+    for (Map.Entry<String, MutableInt> e : ngramCounts.entrySet()) {
+      allGrams[i] = e.getKey();
+      // TODO -- once OPENNLP-1261 is fixed,
+      // change this to e.getValue().getValue().
+      counts[i] = 1;
+      i++;
+    }
+    double[] eval = model.getMaxentModel().eval(allGrams, counts);
+    Language[] arr = new Language[eval.length];
+    for (int j = 0; j < eval.length; j++) {
+      arr[j] = new Language(model.getMaxentModel().getOutcome(j), eval[j]);
+    }
+
+    Arrays.sort(arr, (o1, o2) -> Double.compare(o2.getConfidence(), o1.getConfidence()));
+    return arr;
+  }
+  /**
+   * Size in codepoints at which to chunk the
+   * text for detection.
+   *
+   * @return the chunk size in codepoints
+   */
+  public int getChunkSize() {
+    return chunkSize;
+  }
+
+  /**
+   * Size in codepoints at which to chunk the
+   * text for detection.
+   *
+   * @param chunkSize
+   */
+  public void setChunkSize(int chunkSize) {
+    this.chunkSize = chunkSize;
+  }
+
+  /**
+   * Number of consecutive improvements in the
+   * confidence of the most likely language required
+   * for this language detector to stop.
+   *
+   * @return the minimum consecutive improvements
+   */
+  public int getMinConsecImprovements() {
+    return minConsecImprovements;
+  }
+  /**
+   * Number of consecutive improvements in the
+   * confidence of the most likely language required
+   * for this language detector to stop.
+   *
+   * @param minConsecImprovements minimum consecutive improvements
+   */
+  public void setMinConsecImprovements(int minConsecImprovements) {
+    this.minConsecImprovements = minConsecImprovements;
+  }
+
+  /**
+   * The minimum difference between the highest confidence and the
+   * second highest confidence required to stop.
+   * @return the minimum difference required
+   */
+  public double getMinDiff() {
+    return minDiff;
+  }
+
+  /**
+   * The minimum difference between the highest confidence and the
+   * second highest confidence required to stop.
+   *
+   * Throws {@link IllegalArgumentException} if &lt; 0.0
+   * @param minDiff
+   */
+  public void setMinDiff(double minDiff) {
```

Comment:
It would be better if these are passed directly to the predict method. To allow certain features, such as multi threading, and serializing objects it is better if the component itself is immutable.

Reply:
+1.  Are you ok if I create a config object that can optionally be passed in to `detect`?

Reply:
I think there should be a simple version that kind of works out of the box.

Anyway, we could overload the method with one which takes more parameters, but if there are too many parameters it would be better to pass a config object to it. Or see if we somehow can eliminate some of the parameters.

Reply:
How's it look now? with `probingPredictLanguages(CharSequence txt)`